### PR TITLE
state/themes: Add new reducers, utils, schema, and constants

### PR DESCRIPTION
**Pull request by ockham**
These reducers will later replace the current `themes`, `themesList`, and `themeDetails` ones. This PR doesn't wire them to the live code yet, so their functionality is only covered by tests right now.

I hope the JSDoc blocks for individual reducers are explanatory enough, so I'm not describing them here. If they aren't, just add a comment!

To test: Make sure that the app still works as before (the theme showcase, specifically)

**Edit**: I was actually able to drop the `items` reducer since all information we need is stored inside the `ThemeQueryManager` instances in `queries`.

### CODE REVIEW

File: client/state/themes/test/reducer.js [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+					mood
+				]
+			} );
+
+			expect( state ).to.eql( {
+				'2916284:{"search":"Mood"}': false
+			} );
+		} );
+
+		it( 'should track theme query request failure', () => {
```
Comment (budzanowski) [inline client/state/themes/test/reducer.js]: If the same state is the outcome of two different scenarios I think that we cannot say that the reducers track those scenarios. `false` for both `THEMES_REQUEST_SUCCESS` and `THEMES_REQUEST_FAILURE`

File: client/state/themes/utils.js [hunk 2]
```diff
@@ -4,9 +4,85 @@
...
+ */
+
+/**
+ * Returns a normalized themes query, excluding any values which match the
+ * default theme query.
+ *
+ * @param  {Object} query Themes query
+ * @return {Object}       Normalized themes query
+ */
+export function getNormalizedThemesQuery( query ) {
```
Comment (budzanowski) [inline client/state/themes/utils.js]: `get...` suggests that it is a getter of some sort. I am not an expert but in functional style name of function should explain action performed on its args so maybe `normalizeThemesQuery`?

File: client/state/themes/utils.js [hunk 3]
```diff
@@ -4,9 +4,85 @@
...
+}
+
+/**
+ * Returns a serialized themes query
+ *
+ * @param  {Object} query  Themes query
+ * @param  {Number} siteId Optional site ID
+ * @return {String}        Serialized themes query
+ */
+export function getSerializedThemesQuery( query = {}, siteId ) {
```
Comment (budzanowski) [inline client/state/themes/utils.js]: same as comment above `serializeThemesQuery`?

**Comment by budzanowski** [comment]
Wow nice job - especially tests part. In general :+1: Feel free to address or ignore my nitpicks as they are not very significant.

File: client/state/themes/schema.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
+const themeSchema = {
+	title: 'Theme',
+	type: 'object',
+	properties: {
+		id: {
+			type: 'string'
+		},
```
Comment (seear) [inline client/state/themes/schema.js]: May be more concise if these were on a single lines:
```
id: { type: 'string' },
name: { type: 'string' },
...
```
Unless the linter won't have it.

File: client/state/themes/utils.js [hunk 5]
```diff
@@ -4,9 +4,85 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { DEFAULT_THEME_QUERY } from './constants';
+
+/**
+ * Constants
+ */
+const REGEXP_SERIALIZED_QUERY = /^((\d+):)?(.*)$/;
```
Comment (seear) [inline client/state/themes/utils.js]: The first set of brackets can be a non-capturing group here I think, if you want:
`/^(?:(\d+):)?(.*)$/`

File: client/state/themes/constants.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+/** @ssr-ready **/
+
+export const DEFAULT_THEME_QUERY = {
+	search: '',
+	tier: '',
```
Comment (ockham) [inline client/state/themes/constants.js]: > '' - empty string is equivalent to `ALL`?

Yes. See e.g. [here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/themes/theme-showcase.jsx#L59) (though it's admittedly not very clear from our code).

File: client/state/themes/utils.js (hunk 2 above)
Comment (ockham) [inline client/state/themes/utils.js]: Tricky. A function name that uses a verb (to me, at least) kind of suggests that it transforms an argument, or triggers an action. It's true that the `get` suggests it might be a selector, but I think it's just for lack of synonyms that both use `get`.

Anyway, this is once again code stolen from the `posts` state, so I'd keep the `get`, if only for consistency's sake.

Comment (budzanowski) [inline client/state/themes/utils.js]: `compute` that would be a fun word to use, accurate but feels cliché :)
Argument about consistency makes sense.

**Comment by ockham** [comment]
> I'd really appreciate a walkthrough before we eventually hook it all up, especially around the new query-manager stuff.

Yes, we should definitely do that!

Thanks for reviewing, guys!

File: client/state/themes/test/reducer.js (hunk 1 above)
Comment (ockham) [inline client/state/themes/test/reducer.js]: Hmm. Well essentially, we're covering the `queryRequests` reducer's reaction to different actions here, and the only information it really stores is if a request is ongoing, or not. And both `THEMES_REQUEST_SUCCESS` and `THEMES_REQUEST_FAILURE` signal that the request has finished (as oppsed to ` THEMES_REQUEST`, which we check to set the flag to `true`), so I actually think it kinda makes sense. This is just a consequence of what reducer stores which piece of information when a given action is triggered -- which in case of `queryRequests` clearly isn't a lot 😄 

To put it differently, I wouldn't know how else to cover [this functionality](https://github.com/Automattic/wp-calypso/pull/9476/files#diff-40ab1217977ee8f85be34d691f55b624R77) with tests.
